core/src/main/java/org/bitcoinj/core/Transaction.java

```diff
@@ -1413,67 +1412,64 @@ public synchronized Sha256Hash hashForWitnessSignature(
-                    bosHashOutputs.write(scriptBytes);
-                }
-                hashOutputs = Sha256Hash.hashTwice(bosHashOutputs.toByteArray());
-            } else if (basicSigHashType == SigHash.SINGLE.value && inputIndex < outputs.size()) {
-                ByteArrayOutputStream bosHashOutputs = new ByteArrayOutputStream(256);
-                writeInt64LE(this.outputs.get(inputIndex).getValue().getValue(), bosHashOutputs);
-                byte[] scriptBytes = this.outputs.get(inputIndex).getScriptBytes();
-                bosHashOutputs.write(VarInt.of(scriptBytes.length).serialize());
-                bosHashOutputs.write(scriptBytes);
-                hashOutputs = Sha256Hash.hashTwice(bosHashOutputs.toByteArray());
+        if (signAll) {
+            ByteBuffer bufHashOutputs = ByteBuffer.allocate(this.outputs.stream().mapToInt(
+                    output -> Coin.BYTES + Buffers.lengthPrefixedBytesSize(output.getScriptBytes())
+            ).sum());
+            for (TransactionOutput output : this.outputs) {
+                output.getValue().write(bufHashOutputs);
+                Buffers.writeLengthPrefixedBytes(bufHashOutputs, output.getScriptBytes());
             }
-            writeInt32LE(version, bos);
-            bos.write(hashPrevouts);
-            bos.write(hashSequence);
-            bos.write(inputs.get(inputIndex).getOutpoint().hash().serialize());
-            writeInt32LE(inputs.get(inputIndex).getOutpoint().index(), bos);
-            bos.write(VarInt.of(scriptCode.length).serialize());
-            bos.write(scriptCode);
-            writeInt64LE(prevValue.getValue(), bos);
-            writeInt32LE(inputs.get(inputIndex).getSequenceNumber(), bos);
-            bos.write(hashOutputs);
-            writeInt32LE(this.vLockTime.rawValue(), bos);
-            writeInt32LE(0x000000ff & sigHashType, bos);
-        } catch (IOException e) {
-            throw new RuntimeException(e);  // Cannot happen.
+            hashOutputs = Sha256Hash.wrapReversed(Sha256Hash.hashTwice(bufHashOutputs.array()));
+        } else if (basicSigHashType == SigHash.SINGLE.value && inputIndex < outputs.size()) {
+            TransactionOutput output = this.outputs.get(inputIndex);
+            byte[] scriptBytes = output.getScriptBytes();
+            ByteBuffer bufHashOutputs = ByteBuffer.allocate(Coin.BYTES +
+                    Buffers.lengthPrefixedBytesSize(scriptBytes));
+            output.getValue().write(bufHashOutputs);
+            Buffers.writeLengthPrefixedBytes(bufHashOutputs, scriptBytes);
+            hashOutputs = Sha256Hash.wrapReversed(Sha256Hash.hashTwice(bufHashOutputs.array()));
         }
 
-        return Sha256Hash.twiceOf(bos.toByteArray());
+        ByteBuffer buf = ByteBuffer.allocate(4 + Sha256Hash.LENGTH * 3 + 4 +
+                Buffers.lengthPrefixedBytesSize(scriptCode) + Coin.BYTES + 4 + Sha256Hash.LENGTH + 4 + 4);
```

Comment:
Moving the creation of this buffer up to where the BAOS is allocated would make the diff easier to review and the length calculation could be used on the BAOS version as well.